File: openmp_polybench_refactored/src/metrics.c
#include "metrics.h"
#include <float.h>

// Comparison function for qsort
static int compare_double(const void* a, const void* b) {
    double da = *(const double*)a;
    double db = *(const double*)b;
    if (da < db) return -1;
    if (da > db) return 1;
    return 0;
}

void metrics_init(MetricsCollector* mc, const char* benchmark, const char* dataset, int threads) {
    strncpy(mc->benchmark_name, benchmark, sizeof(mc->benchmark_name) - 1);
    strncpy(mc->dataset, dataset, sizeof(mc->dataset) - 1);
    mc->threads = threads;
    mc->num_results = 0;
    mc->sequential_time_ms = -1.0;  // Not yet set
}

void timing_init(TimingData* td) {
    td->count = 0;
}

void timing_record(TimingData* td, double time_ms) {
    if (td->count < MAX_ITERATIONS) {
        td->times_ms[td->count++] = time_ms;
    }
}

double timing_min(const TimingData* td) {
    if (td->count == 0) return 0.0;
    double min_val = td->times_ms[0];
    for (int i = 1; i < td->count; i++) {
        if (td->times_ms[i] < min_val) min_val = td->times_ms[i];
    }
    return min_val;
}

double timing_max(const TimingData* td) {
    if (td->count == 0) return 0.0;
    double max_val = td->times_ms[0];
    for (int i = 1; i < td->count; i++) {
        if (td->times_ms[i] > max_val) max_val = td->times_ms[i];
    }
    return max_val;
}

double timing_median(const TimingData* td) {
    if (td->count == 0) return 0.0;
    
    double sorted[MAX_ITERATIONS];
    memcpy(sorted, td->times_ms, td->count * sizeof(double));
    qsort(sorted, td->count, sizeof(double), compare_double);
    
    if (td->count % 2 == 0) {
        return (sorted[td->count/2 - 1] + sorted[td->count/2]) / 2.0;
    } else {
        return sorted[td->count/2];
    }
}

double timing_mean(const TimingData* td) {
    if (td->count == 0) return 0.0;
    double sum = 0.0;
    for (int i = 0; i < td->count; i++) {
        sum += td->times_ms[i];
    }
    return sum / td->count;
}

double timing_std(const TimingData* td) {
    if (td->count < 2) return 0.0;
    double mean = timing_mean(td);
    double sum_sq = 0.0;
    for (int i = 0; i < td->count; i++) {
        double diff = td->times_ms[i] - mean;
        sum_sq += diff * diff;
    }
    return sqrt(sum_sq / (td->count - 1));  // Sample standard deviation
}

double compute_speedup(double baseline_ms, double current_ms) {
    if (current_ms <= 0.0) return 1.0;
    return baseline_ms / current_ms;
}

double compute_efficiency(const char* strategy, double speedup, int threads) {
    // Efficiency only meaningful for parallel strategies
    if (!strategy_is_parallel(strategy)) {
        return NAN;  // Not applicable
    }
    if (threads <= 0) return NAN;
    return (speedup / threads) * 100.0;
}

double compute_gflops(double flops, double time_ms) {
    if (time_ms <= 0.0) return 0.0;
    return flops / (time_ms / 1000.0) / 1e9;
}

void metrics_record(MetricsCollector* mc,
                   const char* strategy,
                   const TimingData* timing,
                   double flops,
                   int verified,
                   double max_error) {
    if (mc->num_results >= 32) return;
    
    BenchmarkResult* r = &mc->results[mc->num_results];
    
    // Safe string copy with explicit null termination
    memset(r->benchmark, 0, sizeof(r->benchmark));
    memset(r->dataset, 0, sizeof(r->dataset));
    memset(r->strategy, 0, sizeof(r->strategy));
    strncpy(r->benchmark, mc->benchmark_name, sizeof(r->benchmark) - 1);
    strncpy(r->dataset, mc->dataset, sizeof(r->dataset) - 1);
    strncpy(r->strategy, strategy, sizeof(r->strategy) - 1);
    r->threads = mc->threads;
    r->is_parallel = strategy_is_parallel(strategy);
    
    r->min_ms = timing_min(timing);
    r->median_ms = timing_median(timing);
    r->mean_ms = timing_mean(timing);
    r->std_ms = timing_std(timing);
    
    r->gflops = compute_gflops(flops, r->min_ms);
    
    // Set sequential baseline if this is sequential strategy
    if (strcmp(strategy, "sequential") == 0 || strcmp(strategy, "seq") == 0) {
        mc->sequential_time_ms = r->min_ms;
        r->speedup = 1.0;
    } else if (mc->sequential_time_ms > 0.0) {
        r->speedup = compute_speedup(mc->sequential_time_ms, r->min_ms);
    } else {
        r->speedup = 1.0;  // No baseline yet
    }
    
    r->efficiency_pct = compute_efficiency(strategy, r->speedup, r->threads);
    r->verified = verified;
    r->max_error = max_error;
    r->allocations = 0;  // Not tracked in C
    
    mc->num_results++;
}

void metrics_print_header(void) {
    printf("%-20s %-10s %-8s %-12s %-12s %-10s %-10s %-10s %s\n",
           "Strategy", "Threads", "Parallel", "Min(ms)", "Median(ms)", 
           "GFLOP/s", "Speedup", "Eff(%)", "Verified");
    printf("--------------------------------------------------------------------------------\n");
}

void metrics_print_result(const BenchmarkResult* r) {
    char eff_str[16];
    if (isnan(r->efficiency_pct)) {
        snprintf(eff_str, sizeof(eff_str), "-");
    } else {
        snprintf(eff_str, sizeof(eff_str), "%.1f", r->efficiency_pct);
    }
    
    printf("%-20s %-10d %-8s %-12.3f %-12.3f %-10.2f %-10.2f %-10s %s\n",
           r->strategy,
           r->threads,
           r->is_parallel ? "true" : "false",
           r->min_ms,
           r->median_ms,
           r->gflops,
           r->speedup,
           eff_str,
           r->verified ? "PASS" : "FAIL");
}

void metrics_print_summary(const MetricsCollector* mc) {
    printf("\n");
    printf("=============================================================================\n");
    printf("Benchmark: %s | Dataset: %s | Threads: %d\n", 
           mc->benchmark_name, mc->dataset, mc->threads);
    printf("=============================================================================\n");
    metrics_print_header();
    for (int i = 0; i < mc->num_results; i++) {
        metrics_print_result(&mc->results[i]);
    }
    printf("\n");
}

void metrics_export_csv(const MetricsCollector* mc, const char* filepath) {
    FILE* fp = fopen(filepath, "w");
    if (!fp) {
        fprintf(stderr, "ERROR: Cannot open %s for writing\n", filepath);
        return;
    }
    
    // CSV header - aligned with Julia format
    fprintf(fp, "benchmark,dataset,strategy,threads,is_parallel,min_ms,median_ms,mean_ms,std_ms,gflops,speedup,efficiency_pct,verified,max_error,allocations\n");
    
    for (int i = 0; i < mc->num_results; i++) {
        const BenchmarkResult* r = &mc->results[i];
        
        // Efficiency: empty string for NaN (Julia convention)
        char eff_str[32];
        if (isnan(r->efficiency_pct)) {
            eff_str[0] = '\0';
        } else {
            snprintf(eff_str, sizeof(eff_str), "%.2f", r->efficiency_pct);
        }
        
        fprintf(fp, "%s,%s,%s,%d,%s,%.4f,%.4f,%.4f,%.4f,%.2f,%.2f,%s,%s,%.2e,%ld\n",
                r->benchmark,
                r->dataset,
                r->strategy,
                r->threads,
                r->is_parallel ? "true" : "false",
                r->min_ms,
                r->median_ms,
                r->mean_ms,
                r->std_ms,
                r->gflops,
                r->speedup,
                eff_str,
                r->verified ? "PASS" : "FAIL",
                r->max_error,
                (long)r->allocations);
    }
    
    fclose(fp);
    printf("CSV exported: %s\n", filepath);
}

void metrics_export_json(const MetricsCollector* mc, const char* filepath) {
    FILE* fp = fopen(filepath, "w");
    if (!fp) {
        fprintf(stderr, "ERROR: Cannot open %s for writing\n", filepath);
        return;
    }
    
    char timestamp[64];
    get_timestamp(timestamp, sizeof(timestamp));
    
    fprintf(fp, "{\n");
    fprintf(fp, "  \"metadata\": {\n");
    fprintf(fp, "    \"timestamp\": \"%s\",\n", timestamp);
    fprintf(fp, "    \"benchmark\": \"%s\",\n", mc->benchmark_name);
    fprintf(fp, "    \"dataset\": \"%s\",\n", mc->dataset);
    fprintf(fp, "    \"threads\": %d,\n", mc->threads);
    fprintf(fp, "    \"language\": \"openmp\"\n");
    fprintf(fp, "  },\n");
    fprintf(fp, "  \"results\": [\n");
    
    for (int i = 0; i < mc->num_results; i++) {
        const BenchmarkResult* r = &mc->results[i];
        
        fprintf(fp, "    {\n");
        fprintf(fp, "      \"strategy\": \"%s\",\n", r->strategy);
        fprintf(fp, "      \"threads\": %d,\n", r->threads);
        fprintf(fp, "      \"is_parallel\": %s,\n", r->is_parallel ? "true" : "false");
        fprintf(fp, "      \"min_ms\": %.4f,\n", r->min_ms);
        fprintf(fp, "      \"median_ms\": %.4f,\n", r->median_ms);
        fprintf(fp, "      \"mean_ms\": %.4f,\n", r->mean_ms);
        fprintf(fp, "      \"std_ms\": %.4f,\n", r->std_ms);
        fprintf(fp, "      \"gflops\": %.2f,\n", r->gflops);
        fprintf(fp, "      \"speedup\": %.2f,\n", r->speedup);
        if (isnan(r->efficiency_pct)) {
            fprintf(fp, "      \"efficiency_pct\": null,\n");
        } else {
            fprintf(fp, "      \"efficiency_pct\": %.2f,\n", r->efficiency_pct);
        }
        fprintf(fp, "      \"verified\": %s,\n", r->verified ? "true" : "false");
        fprintf(fp, "      \"max_error\": %.2e\n", r->max_error);
        fprintf(fp, "    }%s\n", (i < mc->num_results - 1) ? "," : "");
    }
    
    fprintf(fp, "  ]\n");
    fprintf(fp, "}\n");
    
    fclose(fp);
    printf("JSON exported: %s\n", filepath);
}

void get_timestamp(char* buf, size_t len) {
    time_t now = time(NULL);
    struct tm* tm_info = localtime(&now);
    strftime(buf, len, "%Y%m%d_%H%M%S", tm_info);
}

File: julia_polybench_refactored/src/common/Metrics.jl
module Metrics
#=
Metrics Collection Module for Julia PolyBench Benchmarks
REFACTORED: Scientifically correct speedup and efficiency formulas

=============================================================================
SPEEDUP AND EFFICIENCY - SCIENTIFIC COMPUTING STANDARDS
=============================================================================

SPEEDUP (S_p):
  S_p = T_1 / T_p
  Where:
    T_1 = execution time with 1 thread (sequential baseline)
    T_p = execution time with p threads/processors
  
  For non-threaded optimizations (BLAS, SIMD):
    S_opt = T_seq / T_opt
    This measures algorithmic/implementation improvement, NOT parallel efficiency

PARALLEL EFFICIENCY (E_p):
  E_p = S_p / p = T_1 / (p * T_p)
  Where:
    S_p = speedup with p threads
    p = number of threads used
  
  Interpretation:
    E_p = 1.0 (100%) = perfect linear scaling
    E_p < 1.0 = sublinear scaling (Amdahl's law, overhead, contention)
    E_p > 1.0 = superlinear scaling (cache effects, rare)

IMPORTANT DISTINCTIONS:
- Parallel efficiency ONLY applies to parallel strategies
- BLAS/SIMD speedup is NOT parallel efficiency - it's algorithmic speedup
- Report BLAS speedup as "speedup" only, not "efficiency"
- Sequential baseline always has efficiency = 100% by definition

AMDAHL'S LAW:
  S_max = 1 / ((1 - f) + f/p)
  Where f = parallelizable fraction
  
=============================================================================
=#

using Printf
using Statistics
using Dates

export BenchmarkResult, MetricsCollector, BenchmarkRunConfig
export record!, print_results, export_csv, export_json
export compute_speedup, compute_parallel_efficiency
export is_parallel_strategy, PARALLEL_STRATEGIES, NON_PARALLEL_STRATEGIES

#=============================================================================
 Strategy Classification
 
 PARALLEL: Uses Julia threads, benefits from more cores
 NON_PARALLEL: Single-threaded optimizations (BLAS handles its own threading)
=============================================================================#
const PARALLEL_STRATEGIES = Set([
    "threads", "threads_static", "threads_dynamic",
    "tiled", "blocked", "tasks", "wavefront",
    "redblack", "parallel"
])

const NON_PARALLEL_STRATEGIES = Set([
    "sequential", "seq", "simd", "blas", "colmajor"
])

function is_parallel_strategy(strategy::AbstractString)::Bool
    return lowercase(String(strategy)) in PARALLEL_STRATEGIES
end

#=============================================================================
 Result Structures
=============================================================================#
struct BenchmarkResult
    benchmark::String          # Kernel name: "2mm", "3mm", etc.
    dataset::String            # Size: "MINI", "SMALL", "MEDIUM", "LARGE", "EXTRALARGE"
    strategy::String           # Strategy name
    threads::Int               # Julia threads available at runtime
    times_ns::Vector{Float64}  # All timing samples in nanoseconds
    allocations::Int           # Memory allocations during kernel
    flops::Float64             # Total FLOPs for the computation
    verified::Bool             # Verification passed
    max_error::Float64         # Maximum verification error
end

# Convenience constructor
function BenchmarkResult(
    benchmark::String,
    dataset::String,
    strategy::String,
    threads::Int,
    times_ns::Vector{Float64},
    allocations::Int,
    flops::Float64;
    verified::Bool=true,
    max_error::Float64=0.0
)
    return BenchmarkResult(
        benchmark, dataset, strategy, threads,
        times_ns, allocations, flops, verified, max_error
    )
end

# Configuration for a benchmark run
struct BenchmarkRunConfig
    warmup_iterations::Int
    timed_iterations::Int
    verify::Bool
    export_csv::Bool
    export_json::Bool
end

BenchmarkRunConfig() = BenchmarkRunConfig(5, 10, true, true, false)

#=============================================================================
 Metrics Collector
=============================================================================#
mutable struct MetricsCollector
    results::Vector{BenchmarkResult}
    timestamp::String
    benchmark_name::String
    dataset::String
    threads::Int
    
    function MetricsCollector(;benchmark::String="", dataset::String="", threads::Int=1)
        new(
            BenchmarkResult[],
            Dates.format(now(), "yyyymmdd_HHMMSS"),
            benchmark,
            dataset,
            threads
        )
    end
end

function record!(mc::MetricsCollector, result::BenchmarkResult)
    push!(mc.results, result)
end

#=============================================================================
 Speedup Calculation
 
 Computes speedup relative to sequential baseline
 S = T_baseline / T_current
=============================================================================#
function compute_speedup(baseline_time_ns::Float64, current_time_ns::Float64)::Float64
    if current_time_ns <= 0.0 || baseline_time_ns <= 0.0
        return 1.0
    end
    return baseline_time_ns / current_time_ns
end

#=============================================================================
 Parallel Efficiency Calculation
 
 ONLY meaningful for parallel strategies
 E = S / p = (T_1 / T_p) / p
 
 Returns:
   - For parallel strategies: (speedup / threads) * 100
   - For non-parallel strategies: NaN (not applicable)
=============================================================================#
function compute_parallel_efficiency(
    strategy::AbstractString,
    speedup::Float64,
    threads::Int
)::Float64
    if !is_parallel_strategy(strategy)
        return NaN  # Efficiency not meaningful for non-parallel strategies
    end
    
    if threads <= 0
        return NaN
    end
    
    return (speedup / threads) * 100.0
end

#=============================================================================
 Print Results - Console Output
=============================================================================#
function print_results(mc::MetricsCollector)
    isempty(mc.results) && return
    
    # Find sequential baseline
    seq_results = filter(r -> lowercase(r.strategy) in ["sequential", "seq"], mc.results)
    seq_time_ns = isempty(seq_results) ? nothing : minimum(seq_results[1].times_ns)
    
    println()
    println("="^100)
    @printf("%-18s | %10s | %10s | %10s | %10s | %8s | %8s | %6s\n",
            "Strategy", "Min(ms)", "Median(ms)", "Mean(ms)", "Std(ms)", "GFLOP/s", "Speedup", "Eff(%)")
    println("="^100)
    
    for r in mc.results
        min_t = minimum(r.times_ns) / 1e6
        med_t = Statistics.median(r.times_ns) / 1e6
        mean_t = Statistics.mean(r.times_ns) / 1e6
        std_t = length(r.times_ns) > 1 ? Statistics.std(r.times_ns) / 1e6 : 0.0
        gflops = r.flops / (minimum(r.times_ns) / 1e9) / 1e9
        
        speedup = seq_time_ns === nothing ? 1.0 : compute_speedup(seq_time_ns, minimum(r.times_ns))
        efficiency = compute_parallel_efficiency(r.strategy, speedup, r.threads)
        
        eff_str = isnan(efficiency) ? "  N/A" : @sprintf("%6.1f", efficiency)
        status = r.verified ? "" : " [FAIL]"
        
        @printf("%-18s | %10.3f | %10.3f | %10.3f | %10.3f | %8.2f | %7.2fx | %s%s\n",
                r.strategy, min_t, med_t, mean_t, std_t, gflops, speedup, eff_str, status)
    end
    println("="^100)
    
    # Legend
    println()
    println("Legend:")
    println("  Speedup = T_sequential / T_strategy")
    println("  Eff(%) = (Speedup / Threads) * 100  [parallel strategies only]")
    println("  N/A = Efficiency not applicable (non-parallel strategy)")
end

#=============================================================================
 CSV Export - For Visualization Pipeline
=============================================================================#
function export_csv(mc::MetricsCollector, filepath::String)
    # Find sequential baseline
    seq_results = filter(r -> lowercase(r.strategy) in ["sequential", "seq"], mc.results)
    seq_time_ns = isempty(seq_results) ? nothing : minimum(seq_results[1].times_ns)
    
    open(filepath, "w") do io
        # Header
        println(io, "benchmark,dataset,strategy,threads,is_parallel,min_ms,median_ms,mean_ms,std_ms,gflops,speedup,efficiency_pct,verified,max_error,allocations")
        
        for r in mc.results
            min_t = minimum(r.times_ns) / 1e6
            med_t = Statistics.median(r.times_ns) / 1e6
            mean_t = Statistics.mean(r.times_ns) / 1e6
            std_t = length(r.times_ns) > 1 ? Statistics.std(r.times_ns) / 1e6 : 0.0
            gflops = r.flops / (minimum(r.times_ns) / 1e9) / 1e9
            
            speedup = seq_time_ns === nothing ? 1.0 : compute_speedup(seq_time_ns, minimum(r.times_ns))
            efficiency = compute_parallel_efficiency(r.strategy, speedup, r.threads)
            is_parallel = is_parallel_strategy(r.strategy)
            
            # Use empty string for NaN efficiency in CSV
            eff_str = isnan(efficiency) ? "" : @sprintf("%.2f", efficiency)
            verified_str = r.verified ? "PASS" : "FAIL"
            
            @printf(io, "%s,%s,%s,%d,%s,%.4f,%.4f,%.4f,%.4f,%.4f,%.4f,%s,%s,%.2e,%d\n",
                    r.benchmark, r.dataset, r.strategy, r.threads,
                    is_parallel, min_t, med_t, mean_t, std_t,
                    gflops, speedup, eff_str, verified_str, r.max_error, r.allocations)
        end
    end
    
    println("CSV exported: $filepath")
end

#=============================================================================
 JSON Export - For Detailed Analysis
=============================================================================#
function export_json(mc::MetricsCollector, filepath::String)
    seq_results = filter(r -> lowercase(r.strategy) in ["sequential", "seq"], mc.results)
    seq_time_ns = isempty(seq_results) ? nothing : minimum(seq_results[1].times_ns)
    
    open(filepath, "w") do io
        println(io, "{")
        println(io, "  \"metadata\": {")
        println(io, "    \"timestamp\": \"$(mc.timestamp)\",")
        println(io, "    \"benchmark\": \"$(mc.benchmark_name)\",")
        println(io, "    \"dataset\": \"$(mc.dataset)\",")
        println(io, "    \"threads\": $(mc.threads),")
        println(io, "    \"julia_version\": \"$(VERSION)\"")
        println(io, "  },")
        println(io, "  \"results\": [")
        
        for (idx, r) in enumerate(mc.results)
            min_t = minimum(r.times_ns) / 1e6
            med_t = Statistics.median(r.times_ns) / 1e6
            mean_t = Statistics.mean(r.times_ns) / 1e6
            std_t = length(r.times_ns) > 1 ? Statistics.std(r.times_ns) / 1e6 : 0.0
            gflops = r.flops / (minimum(r.times_ns) / 1e9) / 1e9
            speedup = seq_time_ns === nothing ? 1.0 : compute_speedup(seq_time_ns, minimum(r.times_ns))
            efficiency = compute_parallel_efficiency(r.strategy, speedup, r.threads)
            
            println(io, "    {")
            println(io, "      \"strategy\": \"$(r.strategy)\",")
            println(io, "      \"is_parallel\": $(is_parallel_strategy(r.strategy)),")
            println(io, "      \"min_ms\": $min_t,")
            println(io, "      \"median_ms\": $med_t,")
            println(io, "      \"mean_ms\": $mean_t,")
            println(io, "      \"std_ms\": $std_t,")
            println(io, "      \"gflops\": $gflops,")
            println(io, "      \"speedup\": $speedup,")
            if !isnan(efficiency)
                println(io, "      \"efficiency_pct\": $efficiency,")
            else
                println(io, "      \"efficiency_pct\": null,")
            end
            println(io, "      \"verified\": $(r.verified),")
            println(io, "      \"max_error\": $(r.max_error),")
            println(io, "      \"allocations\": $(r.allocations),")
            println(io, "      \"all_times_ms\": [$(join(r.times_ns ./ 1e6, ", "))]")
            print(io, "    }")
            println(io, idx < length(mc.results) ? "," : "")
        end
        
        println(io, "  ]")
        println(io, "}")
    end
    
    println("JSON exported: $filepath")
end

end # module Metrics
